### PR TITLE
L1GTPrescaler: implement arbitrary prescales on top of the existing L1 results

### PR DESCRIPTION
**Pull request by fwyzard**
This is for the Legacy GT, I can implement the equivalent functionality for the uGT once the the code base has stabilised.

### CODE REVIEW

**Comment by fwyzard** [comment]
@cmsbuild plase test

**Comment by mulhearn** [comment]
Just to be clear: the uGT emulator does emulate pre-scaling (though I can understand why you might want this)

**Comment by mulhearn** [comment]
+1

**Comment by fwyzard** [comment]
hi @mulhearn ,
yes, I know that the GT and uGT emulators implement prescales, and we do rely on that when working with MC or unbiased data.
This module does something slightly different, which is useful to make tighter L1-accept skims starting from existing L1-accepet data.

File: L1Trigger/GlobalTrigger/plugins/L1GTPrescaler.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  virtual bool filter(edm::Event& event, edm::EventSetup const& setup) override;
+
+  static  void fillDescriptions(edm::ConfigurationDescriptions & descriptions);
+
+private:
+  const edm::EDGetTokenT<L1GlobalTriggerReadoutRecord> m_l1ResultsToken;
+  const std::vector<double> m_algoPrescales;
+  const std::vector<double> m_techPrescales;
+  std::vector<unsigned int> m_algoCounters;
+  std::vector<unsigned int> m_techCounters;
```
Comment (davidlange6) [inline L1Trigger/GlobalTrigger/plugins/L1GTPrescaler.cc]: much safer to make these all std::arrays<..., 128 or 64>

Comment (fwyzard) [inline L1Trigger/GlobalTrigger/plugins/L1GTPrescaler.cc]: hi David,
one issue I have with std::array<T,N> is that it cannot be easily constructed from an std::vector<T>, as returned by config.getParameter<std::vector<double>>(...) :-(

Comment (davidlange6) [inline L1Trigger/GlobalTrigger/plugins/L1GTPrescaler.cc]: Something maybe @Dr15Jones can think of a feature inhancement for. As it is you've assumed that the vector length is correct, so it's a question of a few lines go remove that assumption and to provide a good error message for any incorrect input to the pset

**Comment by fwyzard** [comment]
@davidlange6 I'm fine with either versions, but if you prefer to use std::array more widely, could we have something like
```
edm::ParameterSet::getParameter<std::array<T,N>>(...)
```
?

**Comment by davidlange6** [comment]
> On Feb 1, 2016, at 10:12 PM, Andrea Bocci <notifications@github.com> wrote:
> 
> @davidlange6 I'm fine with either versions, but if you prefer to use std::array more widely, could we have something like
> 
> edm::ParameterSet::getParameter<std::array<T,N>>(...)
> 

is the sort of syntax I was thinking @dr15johns might comment on.

I don’t know the potential limitations.

david

**Comment by davidlange6** [comment]
Please test

**Comment by mulhearn** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Comment by Dr15Jones** [comment]
I added an issue #13180 as a place to have a discussion and to make sure the idea isn't lost.

**Pull request merged**